### PR TITLE
core: sharing operation error handling

### PR DESCRIPTION
**Pull request by tvanderstad**
done: prevent users from performing invalid operations on one device
todo: resolve conflicts that arise when performing sharing operations across devices

### CODE REVIEW

File: core/libs/shared/src/core_tree.rs [hunk 1]
```diff
@@ -1,102 +1,37 @@
...
-use crate::signed_file::SignedFile;
 use crate::tree_like::{Stagable, TreeLike};
 use hmdb::log::SchemaEvent;
 use hmdb::transaction::TransactionTable;
 use std::collections::HashSet;
-use tracing::error;
 use uuid::Uuid;
 
-pub struct CoreTree<'a, 'b, Log>
+impl<'a, F, Log> TreeLike for &mut TransactionTable<'a, Uuid, F, Log>
```
Comment (tvanderstad) [inline core/libs/shared/src/core_tree.rs:8]: putting things back the way they were - impl treelike on transactiontable

File: core/libs/shared/src/core_ops.rs [hunk 2]
```diff
@@ -186,6 +187,10 @@ where
         if self.calculate_deleted(id)? {
             return Err(SharedError::FileNonexistent);
         }
+        match self.access_mode(Owner(account.public_key()), id)? {
```
Comment (tvanderstad) [inline core/libs/shared/src/core_ops.rs:190]: permission check

File: core/libs/shared/src/access_info.rs [hunk 3]
```diff
@@ -25,19 +25,23 @@ pub struct UserAccessInfo {
 
 impl PartialEq for UserAccessInfo {
     fn eq(&self, other: &Self) -> bool {
-        self.encrypted_for == other.encrypted_for && self.encrypted_by == other.encrypted_by
+        self.mode == other.mode
```
Comment (tvanderstad) [inline core/libs/shared/src/access_info.rs:28]: bug fix

File: core/libs/shared/src/access_info.rs [hunk 4]
```diff
@@ -25,19 +25,23 @@ pub struct UserAccessInfo {
...
     pub fn encrypt(
         account: &Account, encrypted_by: &PublicKey, encrypted_for: &PublicKey, key: &AESKey,
+        mode: UserAccessMode,
     ) -> SharedResult<Self> {
         let private_key = account.private_key;
         let user_key = pubkey::get_aes_key(&private_key, encrypted_for)?;
         let encrypted_file_key = symkey::encrypt(&user_key, key)?;
         Ok(UserAccessInfo {
-            mode: UserAccessMode::Owner,
+            mode,
```
Comment (tvanderstad) [inline core/libs/shared/src/access_info.rs:44]: bug fix

File: core/libs/shared/src/lazy.rs [hunk 5]
```diff
@@ -138,13 +156,24 @@ impl<T: Stagable> LazyTree<T> {
         if let Some(name) = self.name.get(id) {
             return Ok(name.clone());
         }
-
-        let key = self.decrypt_key(id, account)?;
-        let name = self.find(id)?.secret_name().to_string(&key)?;
-        self.name.insert(*id, name.clone());
+        let id = if let Some(link) = self.link(id)? { link } else { *id };
```
Comment (tvanderstad) [inline core/libs/shared/src/lazy.rs:159]: when you ask for the name of a file that's linked, you get the link's name instead (relied on by path service in addition to being desired behavior for clients)

File: core/libs/shared/src/lazy.rs [hunk 6]
```diff
@@ -224,30 +256,36 @@ impl<T: Stagable> LazyTree<T> {
...
         self.assert_only_folders_have_children()?;
         self.assert_no_cycles()?;
         self.assert_no_path_conflicts()?;
         // todo
         // self.assert_names_decryptable(account)?;
-        // share validations
+        self.assert_no_shared_links()?;
+        self.assert_no_duplicate_links()?;
+        self.assert_no_broken_links()?;
+        self.assert_no_owned_links()?;
```
Comment (tvanderstad) [inline core/libs/shared/src/lazy.rs:268]: new tree validations for sharing (todo: corresponding merge conflict resolutions)

File: core/tests/sharing_tests.rs [hunk 7]
```diff
@@ -310,81 +302,68 @@ fn delete_pending_share_nonexistent() {
...
+    let folder = core2.create_at_path("shared-folder/").unwrap();
     core2
         .share_file(folder.id, &account1.username, ShareMode::Read)
         .unwrap();
     core2.sync(None).unwrap();
 
     core1.sync(None).unwrap();
     core1
-        .create_link_at_path(&format!("{}/received-folder", &account1.username), folder.id)
+        .create_link_at_path("/received-folder", folder.id)
```
Comment (tvanderstad) [inline core/tests/sharing_tests.rs:318]: updates tests now that paths don't include root

File: core/src/service/share_service.rs [hunk 8]
```diff
@@ -120,6 +127,10 @@ impl RequestContext<'_, '_> {
         if file.owner() == owner {
             return Err(CoreError::FileNotShared);
         }
+        // must have access to file
+        if tree.access_mode(owner, &id)?.is_none() {
+            return Err(CoreError::FileNonexistent); // todo: put this shit everywhere
```
Comment (tvanderstad) [inline core/src/service/share_service.rs:132]: @Parth something to think about - this check probably belongs everywhere the deleted checks are

**Review by Parth — APPROVED** [review]

**Pull request merged**